packages/babel-plugin-transform-unicode-escapes/src/index.js

```diff
@@ -0,0 +1,87 @@
+import { declare } from "@babel/helper-plugin-utils";
+import { types as t } from "@babel/core";
+
+export default declare(api => {
+  api.assertVersion(7);
+
+  const surrogate = /[\ud800-\udfff]/g;
+  const unicodeEscape = /(\\+)u\{([0-9a-fA-F]+)\}/g;
+
+  function escape(code) {
+    let str = code.toString(16);
+    // Sigh, node 6 doesn't have padStart
+    while (str.length < 4) str = "0" + str;
+    return "\\u" + str;
+  }
+  function replacer(match, backslashes, code) {
+    if (backslashes.length % 2 === 0) {
+      return match;
+    }
+
+    const char = String.fromCodePoint(parseInt(code, 16));
+    const escaped = backslashes.slice(0, -1) + escape(char.charCodeAt(0));
+
+    return char.length === 1 ? escaped : escaped + escape(char.charCodeAt(1));
+  }
+  function replaceUnicodeEscapes(str) {
+    return str.replace(unicodeEscape, replacer);
+  }
+
+  return {
+    name: "transform-unicode-escapes",
+    visitor: {
+      Identifier(path) {
+        const { node, key } = path;
+        const { name } = node;
+        const replaced = name.replace(surrogate, escape);
+        if (name === replaced) return;
+
+        const str = t.inherits(t.stringLiteral(name), node);
+
+        if (key === "key") {
+          path.replaceWith(str);
+          return;
+        }
+
+        const { parentPath } = path;
+        if (
+          parentPath.isMemberExpression({ property: node }) ||
+          parentPath.isOptionalMemberExpression({ property: node })
+        ) {
+          parentPath.node.computed = true;
+          path.replaceWith(str);
+          return;
+        }
+
+        throw path.buildCodeFrameError(
+          `Can't represent "${name}" as a bare identifier`,
```

Comment:
When can this happen? Since `𝒜\` is not a valid identifier, isn't it invalid also when represented as `\u{1d49c}`?

Reply:
Per https://www.ecma-international.org/ecma-262/5.1/#sec-7.6, in ES5 the Identifier name aligns to Unicode version 3.0, in which only BMP and PUA Plane are defined, that means 𝒜, introduced later in Unicode version 3.1 and allocated to SMP, was never considered a valid identifier name.

Maybe we can state it clearly that any non-BMP characters are not accepted as identifier name in ES5 and so we cannot transform 𝒜 as an identifier.

Reply:
I'm not sure I understand. `𝒜` is a valid identifier in ES6, and can be represented as both `𝒜` and `\u{1d49c}` in ES6.

```js
var 𝒜 = 1;
console.log(\u{1d49c}); // => 1
```

In ES5, `𝒜` is interpreted as two chars, `\ud835` and `\udc9c` (not the single char `\u{1d49c}`). These individual chars must validate as an identifier. Because individual surrogates don't have Unicode categories, neither is considered a part of `ID Start` or `ID Continue`. So it's not possible to take the bare identifier `𝒜` and output `\ud835\udc9c`, since it'll be invalid.

Reply:
> ECMAScript implementations may recognise identifier characters defined in later editions of the Unicode Standard. If portability is a concern, programmers should only employ identifier characters defined in Unicode 3.0.

`𝒜` was introduced in Unicode 3.1. So `var 𝒜 = 1` may be invalid if the implementation does not align to Unicode >3.0, i.e. Node.js v0.10.

Reply:
@jridgewell Thanks for the explanation, I didn't know that in identifier you must validate the single characters rather than the whole code point.

Reply:
@nicolo-ribaudo lol that is why we have long ugly regex in `@babel/helper-validator-identifier`. And I have thought of replacing that by `\p{ID_Start}\p{ID_Continue}+` but I give up since it will imply different behaviour between Node.js versions.